### ISSUE TITLE
bug in `round(0.005, 2)`

### ISSUE THREAD

**Issue by 80avin-sw** [sympy/sympy#27324]
A bug in round function which makes it deviate from the python's implementation of round and otherwise expected behaviour.

```python
from sympy import S

## sympy's round
S('0.005').round(2)
# 0.0
S('0.5').round()
# 0

## python's round
round(0.005, 2)
# 0.01
round(0.5)
# 1

```

Expected answer is  `S('0.005, 2').round(2) == 0.01` and `S('0.5').round() == 1`, which is what python gives and sympy used to give before the PR https://github.com/sympy/sympy/pull/16608/files
All programming languages ( that I know of ) support this behaviour. I even tried in https://www.geogebra.org/classic and https://www.desmos.com/calculator, they also give same output

This case isn't mentioned in the PR description or conversation, that's why I'm thinking it is an unnoticed bug.

**Comment by asmeurer** [comment]
@smichr

**Comment by LogSunny** [comment]
The issue you're encountering, when rounding `0.005` to 2 decimal places results in `0.0` instead of 0.01, stems from the rounding rule being applied in the method.

You can found `round` method in `sympy/core/expr.py`. 

In the rounding logic, the method is likely using "round half to even" or "bankers' rounding", which is a standard method of rounding when the number is exactly halfway between two possible rounded values. 

Here is a intro to [Bankers' Rounding](https://en.wikipedia.org/wiki/IEEE_754#Roundings_to_nearest).

**Banker's rounding** (also known as **round half to even**) is commonly used in many programming languages and libraries because it helps reduce bias in repeated rounding operations and mitigates cumulative errors that can occur over large datasets. (From ChatGPT)

Hope this helps you.

**Comment by 80avin-sw** [comment]
@LogSunny I didn't know the name "Bankers' Rounding" but I am aware of the fact that such rounding algorithm also exists. Example, wolfram alpha uses the round half to even algorithm.

My arguments
1. The conversation around https://github.com/sympy/sympy/pull/16608 doesn't mention that it is changing from "Round half away from zero" to "Round half to even". Considering that, was it an expected change or an unexpected bug ?
2. Even if it were an expected change, the [Round half away from zero](https://en.wikipedia.org/wiki/Rounding#Rounding_half_away_from_zero) algorithm is used so popularly that sympy should be providing it. This ask becomes stronger when we see that sympy used to provide this algorithm previously.
   Right now, there is nothing in sympy, no function or global parameter, which can provide this behaviour.

**Comment by LogSunny** [comment]
smichr seems to have mentioned in this passage https://github.com/sympy/sympy/pull/16608#issue-430658412 on Apr 9, 2019.

>The round method has been modified to round to even by default in keeping with Python 3 semantics

I think if you use `Sympy` as a mathematical tool, you should pay attention to precision. When you must use **Round Half away from zero**  (for example, to meet the requirements of printing or text output), using Python's built-in decimal library may be your better choice., where you can freely control the rounding method.

**Comment by oscarbenjamin** [comment]
> ```
> ## python's round
> round(0.005, 2)
> # 0.01
> round(0.5)
> # 1
> ```

That is what I see:
```
>>> round(0.005, 2)
0.01
>>> round(0.5)
0
```

**Comment by oscarbenjamin** [comment]
The generic code in Expr.round should not be used for Float, Integer, Rational. Instead we should have something like this:
```python
from sympy import Expr, Rational, Integer, Float, sympify
from fractions import Fraction as F
from decimal import Decimal as D
from mpmath import mpf
import decimal


def sympy_round(e: Expr | float, n=None):
    e = sympify(e)
    if isinstance(e, Integer) and n is None:
        return e
    elif isinstance(e, Rational):
        return round_rational(e, n)
    elif isinstance(e, Float):
        return round_float(e, n)
    else:
        assert False


def round_integer(e: Integer, n=None) -> Integer:
    if n is None or n >= 0:
        return e
    else:
        # 2*e = i*10^k + r, 0 <= r < 10^k
        k = -n
        pow10 = 10**k
        p = e.p
        i, r = divmod(2*p, pow10)
        if r:
            return Integer((i // 2 + 1) * pow10)
        else:
            i, r = divmod(i + 1, 4)
            if r != 3:
                return Integer(2*i*pow10)
            else:
                return Integer((2*i + 1)*pow10)


def round_rational(e: Rational, n: int | None = None) -> Rational:
    if n is None:
        return Integer(_round_rational(e.p, e.q))
    elif n >= 0:
        p10 = 10**n
        return Rational(_round_rational(e.p*p10, e.q), p10)
    else:
        p10 = 10**(-n)
        return Rational(p10 * _round_rational(e.p, e.q*p10))


def _round_rational(p: int, q: int) -> int:
    # 2*e = 2*p/q = i + r/q, 0 <= r/q < 1
    i, r = divmod(2*p, q)
    if r:
        return i // 2 + 1
    else:
        # tiebreaker: round half-even
        i, r = divmod(i + 1, 4)
        if r != 3:
            return 2*i
        else:
            return 2*i + 1


def round_float(e: Float, n=None) -> Integer | Float:
    f = mpf(e._mpf_)
    prec = e._prec
    if n is None:
        return Integer(round(f))
    else:
        return Float(round(f, n), precision=prec)


assert sympy_round(Integer(5), -1) == Integer(0)
assert sympy_round(Integer(6), -1) == Integer(10)
assert sympy_round(Integer(15), -1) == Integer(20)
assert sympy_round(Rational('0.5')) == Integer(0)
assert sympy_round(Rational('1.5')) == Integer(2)
assert sympy_round(Rational('-1.5')) == Integer(-2)
assert sympy_round(Float('0.5')) == Integer('0')
assert sympy_round(Float('1.5')) == Integer('2')
assert sympy_round(Float('-1.5')) == Integer('-2')

assert sympy_round(Rational('-1.5'), 1) == Rational('-1.5')
assert sympy_round(Rational('-1.55'), 1) == Rational('-1.6')
assert sympy_round(Float('-1.5'), 1) == Float('-1.5')
assert sympy_round(Float('-1.55'), 1) == Float('-1.6')

assert round(0.5) == 0
assert round(0.005, 2) == 0.01
```

**Comment by oscarbenjamin** [comment]
Ideally we could defer for rounding to other libraries where possible. I think mpmath should be used for rounding floats and Python's builtin round should be used for integers. Even for rational we could use the stdlibs fraction.Fraction or gmpy2 etc. The Expr.round function should probably return a Round expression that delegates to these other things for basic numbers.

**Comment by 80avin** [comment]
@oscarbenjamin You are correct. My example in the issue description is wrong and is a result of taking shortcuts after I felt I have a convincing argument.

Now, I see that the python's behaviour is because of the floating point approximation and it still follows the round half to even algorithm.
```python
def round_dbg(num, n=None):
   print(f"round({num}, {n}) = round('{num:.20f}', {n}) = {round(num, n)}")

round_dbg(0.005, 2)
# round(0.005, 2) = round('0.00500000000000000010', 2) = 0.01

round_dbg(1.005, 2)
# round(1.005, 2) = round('1.00499999999999989342', 2) = 1.0

# The round half-to-even algorithm can be seen in working here:
round_dbg(0.5)
# round(0.5, None) = round('0.50000000000000000000', None) = 0
round_dbg(1.5)
# round(1.5, None) = round('1.50000000000000000000', None) = 2
```

Also, after learning that the `decimal` module has the `ROUND_HALF_UP` algorithm, I am unblocked.
However, it would still be nice to have such an algorithm in the sympy itself.

**Comment by sylee957** [comment]
I'm unclear about what you mean by:
```
round(0.5)
# 1
```
are you correct that Python rounds 0.5 to 1?
I get the result in pure python:

```
Python 3.13.0 (main, Nov 24 2024, 15:34:11) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> round(0.5)
0
>>> round(0.005)
0
```

**Comment by sylee957** [comment]
> (so, for example, both round(0.5) and round(-0.5) are 0, and round(1.5) is 2)

I think that rounding 0.5 to 0 is right in Python [documentation](https://docs.python.org/3/library/functions.html#round) as well

**Comment by 80avin-sw** [comment]
@sylee957 I've edited the issue description to remove Python's implementation. 

Read: https://github.com/sympy/sympy/issues/27324#issuecomment-2509088552

**Comment by sylee957** [comment]
> However, it would still be nice to have such an algorithm in the sympy itself.

I don't think that there we need to implement the algorithms for rounding in SymPy though.
It would not be performant to use, 
and it would also break the consistency of large part of SymPy because a lot of numeric code was written without any assumption of rounding rules, or at least there is no single place to override the numeric arithmetic in SymPy.

**Comment by 80avin** [comment]
@sylee957 I wasn't asking of replacing now, but providing additionally. 
Nevertheless, if it doesn't make the cut, can you suggest what should we prefer when we want to round arbitray precision floats (sympy Floats) using different rounding methods ?

**Comment by oscarbenjamin** [comment]
It should be possible to use `round` or `.round` with sympy expressions but the implementation that makes it work can come from mpmath or elsewhere.

**Comment by sylee957** [comment]
My opinion is that SymPy don't have implement evaluation for decimals or rounding like `evalf`.
it can be better to generate the evaluation using `lambdify`,
and user can pass their own numeric object like `decimal`, `mpf` to evaluate. There is some demonstration

> Nevertheless, if it doesn't make the cut, can you suggest what should we prefer when we want to round arbitray precision floats (sympy Floats) using different rounding methods ?

I'm not sure if this is something you are looking for

```python
from sympy import *
from decimal import Decimal

x, y = symbols('x y')
f = lambdify([x, y], x + y, modules="math")
```
```
>>> f(Decimal('0.1'), Decimal('0.2'))
Decimal('0.3')
```

I think that there are delicate issues about rounding the result after evaluating mathematical functions, like `sqrt` though.
The issue is that [`decimal`](https://docs.python.org/3/library/decimal.html) for instance, has its own implementation of mathematical functions.
If you want `round` to be consistent, you should use `Decimal.sqrt, Decimal.exp, ...` 
and should not abuse things like `numpy.sqrt` or `math.sqrt` even though they may work for some duck type compatibility.
I think that these things are easier done with code generation though.

The particular issue that I notice, is that there is no symbolic `round` function, even though there are `floor, ceiling`.
We can define one if it is needed. 
I think that evaluation for boundary cases like `0.5` doesn't make sense because there is no universal evaluation rule,
But `round(x).is_integer == True` or `round(round(x)) == round(x)` or `round(x) - x <= 0.5` should hold regardless of rounding rules and can be used to simplify the evaluation.
